test/parallel/test-fs-write-optional-params.js

```diff
@@ -0,0 +1,98 @@
+const buffer = Buffer.from('zyx');
+
+const close = util.promisify(fs.close);
+const open = util.promisify(fs.open);
+const read = util.promisify(fs.read);
+const write = util.promisify(fs.write);
+
+async function testInvalid(fd, expectedCode, ...params) {
+  await assert.rejects(
+    write(fd, ...params),
+    { code: expectedCode }
+  );
+}
+
+async function testValid([ buffer, options ], index) {
+  const dest = path.resolve(tmpdir.path, `rwopt_valid_${index}`);
+  let fd;
+  try {
+    fd = await open(dest, 'w+');
+    const { bytesWritten, buffer: bufferWritten } = await write(fd, buffer, options);
+    const writeBufCopy = Uint8Array.prototype.slice.call(bufferWritten);
+
+    const { bytesRead, buffer: bufferRead } = await read(fd, buffer, options);
+    const readBufCopy = Uint8Array.prototype.slice.call(bufferRead);
+
+    assert.ok(bytesWritten >= bytesRead);
+    if (options.length !== undefined && options.length !== null) {
+      assert.strictEqual(bytesWritten, options.length);
+    }
+    if (options.offset === undefined || options.offset === 0) {
+      assert.deepStrictEqual(writeBufCopy, readBufCopy);
+    }
+    assert.deepStrictEqual(bufferWritten, bufferRead);
+  } finally {
+    if (fd) await close(fd);
+  }
+}
+
+(async () => {
+  const fd = await open(destInvalid, 'w+');
+  // Test if first argument is not wrongly interpreted as ArrayBufferView|string
+  for (const badBuffer of [
+    undefined, null, true, 42, 42n, Symbol('42'), NaN, [], () => {},
+    Promise.resolve(new Uint8Array(1)),
+    {},
+    { buffer: 'amNotParam' },
+    { string: 'amNotParam' },
+    { buffer: new Uint8Array(1).buffer },
+    new Date(),
+    new String('notPrimitive'),
+    { [Symbol.toPrimitive]: (hint) => 'amObject' },
+
+    // TODO(LiviaMedeiros): add the following after DEP0162 EOL
+    // { toString() { return 'amObject'; } },
+  ]) {
+    await testInvalid(fd, 'ERR_INVALID_ARG_TYPE', badBuffer, {});
+  }
+
+  // First argument (buffer or string) is mandatory
+  await testInvalid(fd, 'ERR_INVALID_ARG_TYPE');
+
+  // Various invalid options
+  await testInvalid(fd, 'ERR_OUT_OF_RANGE', buffer, { length: 5 });
+  await testInvalid(fd, 'ERR_OUT_OF_RANGE', buffer, { offset: 5 });
+  await testInvalid(fd, 'ERR_OUT_OF_RANGE', buffer, { length: 1, offset: 3 });
+  await testInvalid(fd, 'ERR_OUT_OF_RANGE', buffer, { length: -1 });
+  await testInvalid(fd, 'ERR_OUT_OF_RANGE', buffer, { offset: -1 });
+  await testInvalid(fd, 'ERR_INVALID_ARG_TYPE', buffer, { offset: false });
+  await testInvalid(fd, 'ERR_INVALID_ARG_TYPE', buffer, { offset: true });
+
+  // Test compatibility with fs.read counterpart
+  [
+    [ buffer, {} ],
+    [ buffer, { length: 1 } ],
+    [ buffer, { position: 5 } ],
+    [ buffer, { length: 1, position: 5 } ],
+    [ buffer, { length: 1, position: -1, offset: 2 } ],
+    [ buffer, { length: null } ],
+    [ buffer, { position: null } ],
+    [ buffer, { offset: 1 } ],
+  ].forEach(await testValid);
```

Comment:
You need to `Promise.all` that
```suggestion
  ].map(testValid);
```

Reply:
Whoops :smile: